packages/python-packages/apiview-stub-generator/apistub/_stub_generator.py

```diff
@@ -136,22 +136,34 @@ def install_extra_dependencies(self):
                 logging.info(f"Failed to install extra dependency: {extra}")
                 pass
 
-    def generate_tokens(self):
-        # Extract package to temp directory if it is wheel or sdist
-        if self.pkg_path.endswith(".whl") or self.pkg_path.endswith(".zip"):
-            pkg_root_path = self.wheel_path
-            pkg_name, version = self._parse_pkg_name()
-            namespace = self.get_module_root_name(pkg_root_path)
+    def _get_pkg_metadata(self):
+        metadata = get_metadata(self.pkg_path)
+        pkg_name = metadata.name
+        version = metadata.version
+        pkg_root_path = self.wheel_path or self.pkg_path
+        # TODO: pyproject.toml + whl/sdist: metadata.extras_require doesn't pick up optional dependencies
+        self.extras_require = metadata.provides_extras
+        if self.wheel_path:
+            namespace = self._get_whl_root_namespace(pkg_root_path)
         else:
-            # package root is passed as arg to parse
-            pkg_root_path = self.pkg_path
-            pkg_name, version, namespace, self.extras_require = parse_setup_py(pkg_root_path)
+            namespace = self._get_package_root_namespace(pkg_root_path)
+        return pkg_root_path, pkg_name, version, namespace
 
-        logging.debug("package name: {0}, version:{1}, namespace:{2}".format(pkg_name, version, namespace))
+    def _get_package_root_namespace(self, pkg_root_path):
+        for root, dirs, files in os.walk(pkg_root_path):
```

Comment:
[Replacing - with .](https://github.com/Azure/azure-sdk-for-python/blob/eec0113e48c26a1f416203ba6f82c233895ec8ef/tools/azure-sdk-tools/ci_tools/parsing/parse_functions.py#L393) doesn't work for packages like `azure-eventhub-checkpointstoreblob` where the namespace includes an extra path : `azure.eventhub.extensions.checkpointstoreblob.

Reply:
I believe this is causing the following failure, so we can't do this: `Skipping module build.lib.apistubgentest. Module should start with mnt.vss._work.1.s.packages.python-packages.apiview-stub-generator....apistubgentest`

I don't see an alternative to parsing pyproject.toml/setup.py. Thoughts?

Reply:
discussion w Laurent: Can we avoid passing in namespace here? Can't we just pick up namespace as we go during parsing?

Reply:
> [Replacing - with .](https://github.com/Azure/azure-sdk-for-python/blob/eec0113e48c26a1f416203ba6f82c233895ec8ef/tools/azure-sdk-tools/ci_tools/parsing/parse_functions.py#L393) doesn't work for packages like azure-eventhub-checkpointstoreblob where the namespace doesn't follow the package name

Yeah. This is about how I would do it @swathipil . Unless the pkginfo metadata contains the top-level namespace?

Reply:
I do think this might traverse a bit too deep? If you absolutely need the namespace I think the first folder deeper than the LAST [pure extension dunder init](https://github.com/Azure/azure-sdk-for-python/blob/770bb3c003c348898e45869298d3e3bd48e21ff6/sdk/core/azure-core/azure/__init__.py#L1) is the one you want.

Reply:
V helpful, thanks @scbedd !

Reply:
I am going to add this exact concept to `parse_pyproject.toml` for discovering the root namespace of a package in azure-sdk-tools. Soon ™️